subsys/net/lib/sntp/sntp_simple.c

```diff
@@ -41,6 +41,16 @@ static int sntp_simple_helper(struct sockaddr *addr, socklen_t addr_len, uint32_
 		if (iter_timeout < 1000) {
 			iter_timeout *= 2;
 		}
+
+		/* The context needs to be reopened (to have a new source port in the UDP packet)
```

Comment:
I am not convinced this is needed, isn't the end result the same if we hit a timeout in one iteration, the packet is dropped in both cases?

Reply:
Your diagram says we get `-EINVAL` for the 2nd received packet, what about if we check and handle that properly (meaning we check if we had a timeout earlier and now got EINVAL error) and do not exit the loop in `sntp_simple_helper()` in this case. Then we would not need to open/close the socket.

Reply:
I agree that your proposal is a better approach because it handles system resources more efficiently.

I am worried that responding to such a generic errno such as `-EINVAL` will be problematic however because `errno` values from socket operations (I see zsock_poll and zsock_recv `errno` values being returned directly) may return that value and we would incorrectly assume that the packet was successfully received but the timestamp was mismatched.

I would note that I probably don't have enough system knowledge to make an accurate judgement call on that possibility though.

Another possible solution I considered was making the initial timeout a Kconfig value w/default to the current but decided on the socket close/open because it addresses the core problem rather than addressing the symptom the problem causes.

Reply:
Hi @jukkar, should an issue be made to track the discussion of this PR? How should I proceed to have this bug fixed?

Reply:
> I am worried that responding to such a generic errno such as `-EINVAL` will be problematic however because `errno` values from socket operations (I see zsock_poll and zsock_recv `errno` values being returned directly) may return that value and we would incorrectly assume that the packet was successfully received with the timestamp was mismatched.

We could return a different errno than `-EINVAL` in this case, for example `-ERANGE` could be used. This way we could detect the issue and not get confused with any socket return code.

Reply:
I was writing the code to test this solution when I realized that this also still doesn't necessarily work. We would need to expose `sntp_recv_response` from sntp.c.

The problem is that if we get this -ERANGE error from `sntp_query` and we know that we are looping in the >first iteration of the loop, then we need to call `sntp_recv_response` again, not call `sntp_query` again. Calling query will generate a new request packet and invalidate the context for the next received response. I will write up something that does this but I am making you aware of it earlier than it'll be ready.